### ISSUE TITLE
Support generating interface files separately (and before) compiled code

### ISSUE THREAD

**Issue by DemiMarie** [rust-lang/rust#39541]
Rustc generates files containing both interface information (needed to compile other crates) and compiled object code (needed for linking only).  Currently, any multi-crate build cannot start compiling other crates until both are available, since they are combined into one file.

However, there is no fundamental need for other crates to wait on the compiled object code.  They really only need to wait on the interface information.  This issue covers making the interface information available sooner, and providing a way to notify another process (perhaps by writing to a file descriptor/handle) when it is available.

**Comment by KalitaAlexey** [comment]
@DemiMarie,
Why?
Compilation uses as much resources as possible.
There will be no win with the feature.

**Comment by est31** [comment]
I think I've suggested something like this before: https://internals.rust-lang.org/t/pre-rfc-generate-headers-for-greater-parallelism/4523

**Comment by Ixrec** [comment]
As mentioned on that thread, the potential benefits of this idea will probably fall out of https://github.com/rust-lang/rust/issues/38913 and incremental compilation or at least have to wait until both of those make more progress.

**Comment by Ericson2314** [comment]
This is good for doing a "proof by construction" public-in-private check.

**Comment by steveklabnik** [comment]
Triage: there's been some work on some stuff like this recently, I think?

**Comment by est31** [comment]
@steveklabnik Yeah I'd say this issue is resolved in favour of #60988

**Comment by steveklabnik** [comment]
Ah yes, that's the issue, thank you for finding it!

If anyone objects, please let me know, but I believe this is a duplicate of that.
